### PR TITLE
IZE-378 ize toml schema

### PR DESCRIPTION
**Pull request by psihachina**
## Changelog:
- add json schema for `ize` configuration
- describe the configuration structure
- add schema package to validate the configuration
- add command to test the schema validation

## Test:
[asciicast](https://asciinema.org/a/k7FDkytEak9ydddfr9bqkjowO)

### CODE REVIEW

File: internal/commands/ize.go [hunk 1]
```diff
@@ -87,11 +88,14 @@ func init() {
 	rootCmd.PersistentFlags().String("terraform-version", "", "set terraform-version")
 	rootCmd.PersistentFlags().String("prefer-runtime", "native", "set prefer runtime (native or docker)")
 	rootCmd.Flags().StringP("tag", "t", "", "set tag")
-	viper.BindPFlags(rootCmd.PersistentFlags())
+	// viper.BindPFlags(rootCmd.PersistentFlags())
 	rootCmd.PersistentFlags().VisitAll(func(f *pflag.Flag) {
-		viper.BindPFlag(strings.ReplaceAll(f.Name, "_", "-"), rootCmd.PersistentFlags().Lookup(f.Name))
+		viper.BindPFlag(strings.ReplaceAll(f.Name, "-", "_"), rootCmd.PersistentFlags().Lookup(f.Name))
```
Comment (AutomationD) [inline internal/commands/ize.go:93]: @psihachina Can you please give more info on why this change is required?
What tests have we done to validate that ENV, flags and configs are still working?

File: internal/config/config.go [hunk 2]
```diff
@@ -160,7 +160,7 @@ func GetConfig() (*Config, error) {
 func InitConfig() {
 	viper.SetEnvPrefix("IZE")
 	viper.AutomaticEnv()
-	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	viper.SetEnvKeyReplacer(strings.NewReplacer("_", "-"))
```
Comment (AutomationD) [inline internal/config/config.go:163]: @psihachina Can you please give more info on why this change is required?
What tests have we done to validate that ENV, flags and configs are still working?

File: internal/commands/ize.go [hunk 3]
```diff
@@ -87,11 +88,14 @@ func init() {
...
 	})
 
+	// [log-level aws-profile projects_path env_dir terraform-version prefer_runtime prefer-runtime aws-region home infra_dir root_dir config-file plain-text]
```
Comment (AutomationD) [inline internal/commands/ize.go:96]: @psihachina Can you please give more info on what's happening here?

File: internal/config/project.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package config
+
+type Project struct {
```
Comment (AutomationD) [inline internal/config/project.go:3]: @psihachina What's "Project" in this case? Is it the whole monorepo? If yes, I would avoid naming it as Project, as at the moment project is something that is in the projects/ directory, so either we need to rename those or come up with a name of the "thing" that happens in the ize-structured repo.

...Maybe we should rename the projects/ directory to apps/ and then a project would be logical? 

Let me know what you think.

File: internal/schema/ize-spec.json [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        "namespace": {
+            "type": "string"
+        },
+        "plain_text": {
+            "type": "boolean"
+        },
+        "prefer_runtime": {
+            "type": "string"
+        },
+        "projects_path": {
```
Comment (AutomationD) [inline internal/schema/ize-spec.json:38]: That's the path I'm referring to, it's quite confusing to have the repo as a project and projects path. Maybe this should be come "apps_path"?

File: internal/schema/ize-spec.json [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                        "root_domain_name": {
+                            "type": "string"
+                        },
+                        "aws_region": {
+                            "type": "string"
+                        },
+                        "aws_profile": {
+                            "type": "string"
+                        },
+                        "Tunnel": {
```
Comment (AutomationD) [inline internal/schema/ize-spec.json:81]: Is this supposed to be capitalized? Why?

File: internal/commands/validate/validate.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+		Short: "Validate configuration (only for test)",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.SilenceUsage = true
+			err := schema.Validate(viper.AllSettings())
+
+			if err != nil {
+				return err
+			}
+
+			pterm.Success.Println("All good!")
```
Comment (AutomationD) [inline internal/commands/validate/validate.go:22]:
```suggestion
			pterm.Success.Println("Config structure, env vars and flags look valid. ")
```

File: internal/schema/schema.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/xeipuuv/gojsonschema"
+
+	// Enable support for embedded static resources
+	_ "embed"
+)
+
+//go:embed ize-spec.json
```
Comment (AutomationD) [inline internal/schema/schema.go:13]: 👍🏻

**Review by AutomationD — CHANGES_REQUESTED** [review]
Great start. I have some comments, let's discuss if needed.

File: internal/schema/schema.go (hunk 8 above)
Comment (psihachina) [reply internal/schema/schema.go:13]: [❓](https://emojipedia.org/question-mark/)

File: internal/config/project.go (hunk 4 above)
Comment (psihachina) [reply internal/config/project.go:3]: Project = monorepo. Also you have a created task for rename `projects_dir` to `apps_dir`. Is there any point in discussing it then?

File: internal/commands/ize.go (hunk 1 above)
Comment (psihachina) [reply internal/commands/ize.go:93]: @AutomationD 
It was wrong. We need the flags `-` to be replaced with `_`. Otherwise, our flags are not bound to the identical value with `_` and are duplicated

For examples:
Old output all settings:
`
map[app:map[goblin:map[path:./projects/goblin type:ecs] squibby:map[path:./projects/squibby type:ecs]] apps_path:/home/psih/go/src/github.com/psihachina/ize/examples/ecs-apps-monorepo/projects aws-profile:test aws-region: aws_profile:default aws_region:us-east-1 config-file: env:test-ecs-apps env_dir:/home/psih/go/src/github.com/psihachina/ize/examples/ecs-apps-monorepo/.ize/env/test-ecs-apps home:/home/psih infra_dir:/home/psih/go/src/github.com/psihachina/ize/examples/ecs-apps-monorepo/.ize log-level: namespace:testnut plain-text:false prefer-runtime:native prefer_runtime:native root_dir:/home/psih/go/src/github.com/psihachina/ize/examples/ecs-apps-monorepo tag:de6c646 terraform-version: terraform_version:1.1.7 tf_log:]
`

With this change:
`
map[app:map[goblin:map[path:./projects/goblin type:ecs] squibby:map[path:./projects/squibby type:ecs]] aws_profile:test aws_region:us-east-1 config_file: env:test-ecs-apps env_dir:/home/psih/go/src/github.com/psihachina/ize/examples/ecs-apps-monorepo/.ize/env/test-ecs-apps home:/home/psih infra_dir:/home/psih/go/src/github.com/psihachina/ize/examples/ecs-apps-monorepo/.ize log_level: namespace:testnut plain_text:false prefer_runtime:native projects_path:/home/psih/go/src/github.com/psihachina/ize/examples/ecs-apps-monorepo/projects root_dir:/home/psih/go/src/github.com/psihachina/ize/examples/ecs-apps-monorepo tag:de6c646 terraform_version:1.1.7 tf_log:]
`

File: internal/config/config.go (hunk 2 above)
Comment (psihachina) [reply internal/config/config.go:163]: @AutomationD 
I don't think it's necessary at all

Please check out these examples:
https://github.com/spf13/viper/blob/0add84cdf25b13a28c405bf03ab1acfa9b35c5db/viper_test.go#L851
https://github.com/spf13/viper/blob/0add84cdf25b13a28c405bf03ab1acfa9b35c5db/viper_test.go#L713

Comment (AutomationD) [reply internal/config/config.go:163]: Let's remove it then

File: internal/commands/ize.go (hunk 3 above)
Comment (AutomationD) [reply internal/commands/ize.go:96]: Let's remove

**Review by AutomationD — APPROVED** [review]
